File: CaravanCMS.Api/Services/FuzzyMatcher.cs
using CaravanCMS.Api.Data;
using CaravanCMS.Core;
using FuzzySharp;
using Microsoft.EntityFrameworkCore;

namespace CaravanCMS.Api.Services;

/// <summary>
/// Matches files to caravans using a tiered confidence strategy:
/// 1. VIN found in filename (0.95 confidence)
/// 2. Registration found in filename (0.88 confidence)
/// 3. VIN or reg found in folder path (0.72 confidence)
/// 3.5. Same vehicle folder as an already-linked document (0.68 confidence)
/// 4. Fuzzy make/model match against filename tokens (0.35–0.60 confidence)
///
/// Tier 3.5 also handles VIN→rego inference: when a vehicle folder is named by VIN
/// and other docs in that folder are already linked to a rego, new files inherit it.
/// </summary>
public class FuzzyMatcher
{
    private readonly ApplicationDbContext _db;
    private readonly ILogger<FuzzyMatcher> _logger;

    // Minimum fuzzy ratio (0–100) for a make/model match to be considered meaningful
    private const int MinFuzzyScore = 65;

    public FuzzyMatcher(ApplicationDbContext db, ILogger<FuzzyMatcher> logger)
    {
        _db = db;
        _logger = logger;
    }

    /// <summary>
    /// Attempts to match a file to a caravan. Returns the best match found.
    /// The FileMatchDto.IsMatched is true when Confidence >= 0.5.
    /// </summary>
    /// <param name="filePath">Absolute path to the file to match.</param>
    /// <param name="caravanCache">Pre-loaded caravan list; queried from DB if null.</param>
    /// <param name="folderRegoIndex">
    ///   Maps directory paths to a single RegistrationNumber, built from already-linked documents.
    ///   Used for Tier 3.5 same-folder inference. Pass null to skip this tier.
    /// </param>
    public async Task<FileMatchDto> MatchFileAsync(
        string filePath,
        List<Caravan>? caravanCache = null,
        IReadOnlyDictionary<string, string>? folderRegoIndex = null)
    {
        caravanCache ??= await _db.Caravans.AsNoTracking().ToListAsync();

        string fileName = Path.GetFileNameWithoutExtension(filePath).ToUpper();
        string folderPath = (Path.GetDirectoryName(filePath) ?? string.Empty).ToUpper();
        string fileNameWithExt = Path.GetFileName(filePath);

        // Check if already linked in DB
        Document? existingLink = await _db.Documents.FirstOrDefaultAsync(d => d.FilePath == filePath);
        if (existingLink is not null)
        {
            Caravan? linked = caravanCache.FirstOrDefault(c => c.RegistrationNumber == existingLink.RegistrationNumber);
            return new FileMatchDto
            {
                FilePath = filePath,
                FileName = fileNameWithExt,
                FileExtension = Path.GetExtension(filePath).ToLowerInvariant(),
                FileSizeBytes = new FileInfo(filePath).Length,
                Confidence = 1.0,
                MatchMethod = "AlreadyLinked",
                SuggestedRegistrationNumber = existingLink.RegistrationNumber,
                SuggestedCaravanDescription = linked is not null ? DescribeCaravan(linked) : null,
                LinkedRegistrationNumber = existingLink.RegistrationNumber,
                AlreadyLinked = true
            };
        }

        FileMatchDto best = new()
        {
            FilePath = filePath,
            FileName = fileNameWithExt,
            FileExtension = Path.GetExtension(filePath).ToLowerInvariant(),
            FileSizeBytes = new FileInfo(filePath).Length
        };

        foreach (Caravan caravan in caravanCache)
        {
            TryMatch(caravan, fileName, folderPath, best);
        }

        // ── Tier 3.5: Same vehicle-folder as an already-linked document ────────
        // Walk up the directory tree. If we find a folder where all linked docs share
        // one rego (and our current best is weaker), inherit that rego.
        // This handles gas certs, and folders named by VIN rather than rego.
        if (folderRegoIndex is not null && best.Confidence < 0.68)
        {
            string? dir = Path.GetDirectoryName(filePath);
            for (int depth = 0; depth < 5 && dir is not null; depth++, dir = Path.GetDirectoryName(dir))
            {
                if (folderRegoIndex.TryGetValue(dir, out string? inferredRego))
                {
                    Caravan? c = caravanCache.FirstOrDefault(x => x.RegistrationNumber == inferredRego);
                    double conf = depth <= 1 ? 0.68 : 0.62;
                    if (conf > best.Confidence)
                    {
                        best.SuggestedRegistrationNumber = inferredRego;
                        best.SuggestedCaravanDescription = c is not null ? DescribeCaravan(c) : inferredRego;
                        best.Confidence = conf;
                        best.MatchMethod = "SameFolderInference";
                    }
                    break;
                }
            }
        }

        if (best.Confidence < 0.3)
            _logger.LogDebug("No match found for: {File}", fileNameWithExt);
        else
            _logger.LogDebug("Matched {File} → caravan {Rego} via {Method} ({Score:P0})",
                fileNameWithExt, best.SuggestedRegistrationNumber, best.MatchMethod, best.Confidence);

        return best;
    }

    /// <summary>
    /// Builds a directory→RegistrationNumber lookup from already-linked document paths.
    /// A directory is included only when all linked docs under it share the same single rego —
    /// ensuring the inference is unambiguous.
    /// </summary>
    public static Dictionary<string, string> BuildFolderRegoIndex(
        IEnumerable<(string FilePath, string Rego)> linkedDocs)
    {
        // Accumulate all regos seen under each ancestor directory
        var folderRegos = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);

        foreach ((string filePath, string rego) in linkedDocs)
        {
            string? dir = Path.GetDirectoryName(filePath);
            for (int depth = 0; depth < 6 && dir is not null; depth++, dir = Path.GetDirectoryName(dir))
            {
                if (!folderRegos.TryGetValue(dir, out HashSet<string>? regos))
                    folderRegos[dir] = regos = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                regos.Add(rego);
            }
        }

        // Only keep unambiguous entries (one rego per folder)
        return folderRegos
            .Where(kv => kv.Value.Count == 1)
            .ToDictionary(kv => kv.Key, kv => kv.Value.First(), StringComparer.OrdinalIgnoreCase);
    }

    private static void TryMatch(Caravan caravan, string fileName, string folderPath, FileMatchDto best)
    {
        // ── Tier 1: VIN in filename (highest confidence) ──────────────────────
        if (!string.IsNullOrEmpty(caravan.Vin) && caravan.Vin.Length >= 5)
        {
            string vin = caravan.Vin.ToUpper();
            if (fileName.Contains(vin))
            {
                if (0.95 > best.Confidence)
                    Update(best, caravan, 0.95, "VinInFilename");
                return;
            }
        }

        // ── Tier 2: Registration in filename ─────────────────────────────────
        if (!string.IsNullOrEmpty(caravan.RegistrationNumber) && caravan.RegistrationNumber.Length >= 3)
        {
            string reg = caravan.RegistrationNumber.ToUpper().Replace(" ", "");
            string fileNameClean = fileName.Replace(" ", "").Replace("-", "").Replace("_", "");

            if (fileNameClean.Contains(reg))
            {
                if (0.88 > best.Confidence)
                    Update(best, caravan, 0.88, "RegInFilename");
                return;
            }
        }

        // ── Tier 3: VIN or reg in folder path ────────────────────────────────
        if (!string.IsNullOrEmpty(caravan.Vin) && caravan.Vin.Length >= 5 &&
            folderPath.Contains(caravan.Vin.ToUpper()))
        {
            if (0.72 > best.Confidence)
                Update(best, caravan, 0.72, "VinInFolderPath");
            return;
        }

        if (!string.IsNullOrEmpty(caravan.RegistrationNumber) && caravan.RegistrationNumber.Length >= 3 &&
            folderPath.Contains(caravan.RegistrationNumber.ToUpper().Replace(" ", "")))
        {
            if (0.72 > best.Confidence)
                Update(best, caravan, 0.72, "RegInFolderPath");
            return;
        }

        // ── Tier 4: Fuzzy make/model match ────────────────────────────────────
        if (!string.IsNullOrEmpty(caravan.Make) && !string.IsNullOrEmpty(caravan.Model))
        {
            string target = $"{caravan.Make} {caravan.Model}".ToUpper();
            int score = Fuzz.PartialRatio(fileName, target);

            if (score >= MinFuzzyScore)
            {
                // Normalise FuzzySharp 0-100 score into 0.35–0.60 range
                double confidence = 0.35 + ((score - MinFuzzyScore) / (double)(100 - MinFuzzyScore)) * 0.25;
                if (confidence > best.Confidence)
                    Update(best, caravan, confidence, "FuzzyMakeModel");
            }
        }
    }

    private static void Update(FileMatchDto match, Caravan caravan, double confidence, string method)
    {
        match.SuggestedRegistrationNumber = caravan.RegistrationNumber;
        match.SuggestedCaravanDescription = DescribeCaravan(caravan);
        match.Confidence = confidence;
        match.MatchMethod = method;
    }

    private static string DescribeCaravan(Caravan c) =>
        $"{c.Year} {c.Make} {c.Model} — {c.RegistrationNumber ?? c.Vin ?? "No ID"}".Trim();
}

File: CaravanCMS.Worker/scripts/lib/fuzzyMatcher.ts
import * as path from "node:path";
import * as fs from "node:fs";
import { partial_ratio } from "fuzzball";

/**
 * Port of CaravanCMS.Api/Services/FuzzyMatcher.cs — used only by the
 * one-time local→R2 file migration script (scripts/migrate-files.ts), never
 * by the live Worker (Workers has no local filesystem to scan at runtime).
 * Tiered confidence strategy, thresholds and formula kept identical to the
 * C# original so migrated links match what the old system would have chosen:
 *  1. VIN in filename (0.95)
 *  2. Registration in filename (0.88)
 *  3. VIN or reg in folder path (0.72)
 *  3.5. Same vehicle folder as an already-linked document (0.68 / 0.62)
 *  4. Fuzzy make/model match (0.35–0.60), fuzzball.partial_ratio ports
 *     FuzzySharp's Fuzz.PartialRatio (both descend from fuzzywuzzy).
 */

const MIN_FUZZY_SCORE = 65;

export interface CaravanForMatching {
  registrationNumber: string;
  vin: string | null;
  make: string | null;
  model: string | null;
  year: number | null;
}

export interface FileMatch {
  filePath: string;
  fileName: string;
  fileExtension: string;
  fileSizeBytes: number;
  confidence: number;
  matchMethod: string;
  suggestedRegistrationNumber: string | null;
  suggestedCaravanDescription: string | null;
  alreadyLinked: boolean;
}

function describeCaravan(c: CaravanForMatching): string {
  return `${c.year ?? ""} ${c.make ?? ""} ${c.model ?? ""} — ${c.registrationNumber || c.vin || "No ID"}`.trim();
}

function tryMatch(
  caravan: CaravanForMatching,
  fileName: string,
  folderPath: string,
  best: { confidence: number; suggestedRegistrationNumber: string | null; suggestedCaravanDescription: string | null; matchMethod: string },
): void {
  // Tier 1: VIN in filename
  if (caravan.vin && caravan.vin.length >= 5) {
    const vin = caravan.vin.toUpperCase();
    if (fileName.includes(vin)) {
      if (0.95 > best.confidence) {
        best.suggestedRegistrationNumber = caravan.registrationNumber;
        best.suggestedCaravanDescription = describeCaravan(caravan);
        best.confidence = 0.95;
        best.matchMethod = "VinInFilename";
      }
      return;
    }
  }

  // Tier 2: Registration in filename
  if (caravan.registrationNumber && caravan.registrationNumber.length >= 3) {
    const reg = caravan.registrationNumber.toUpperCase().replaceAll(" ", "");
    const fileNameClean = fileName.replaceAll(" ", "").replaceAll("-", "").replaceAll("_", "");
    if (fileNameClean.includes(reg)) {
      if (0.88 > best.confidence) {
        best.suggestedRegistrationNumber = caravan.registrationNumber;
        best.suggestedCaravanDescription = describeCaravan(caravan);
        best.confidence = 0.88;
        best.matchMethod = "RegInFilename";
      }
      return;
    }
  }

  // Tier 3: VIN or reg in folder path
  if (caravan.vin && caravan.vin.length >= 5 && folderPath.includes(caravan.vin.toUpperCase())) {
    if (0.72 > best.confidence) {
      best.suggestedRegistrationNumber = caravan.registrationNumber;
      best.suggestedCaravanDescription = describeCaravan(caravan);
      best.confidence = 0.72;
      best.matchMethod = "VinInFolderPath";
    }
    return;
  }
  if (
    caravan.registrationNumber &&
    caravan.registrationNumber.length >= 3 &&
    folderPath.includes(caravan.registrationNumber.toUpperCase().replaceAll(" ", ""))
  ) {
    if (0.72 > best.confidence) {
      best.suggestedRegistrationNumber = caravan.registrationNumber;
      best.suggestedCaravanDescription = describeCaravan(caravan);
      best.confidence = 0.72;
      best.matchMethod = "RegInFolderPath";
    }
    return;
  }

  // Tier 4: Fuzzy make/model
  if (caravan.make && caravan.model) {
    const target = `${caravan.make} ${caravan.model}`.toUpperCase();
    const score = partial_ratio(fileName, target);
    if (score >= MIN_FUZZY_SCORE) {
      const confidence = 0.35 + ((score - MIN_FUZZY_SCORE) / (100 - MIN_FUZZY_SCORE)) * 0.25;
      if (confidence > best.confidence) {
        best.suggestedRegistrationNumber = caravan.registrationNumber;
        best.suggestedCaravanDescription = describeCaravan(caravan);
        best.confidence = confidence;
        best.matchMethod = "FuzzyMakeModel";
      }
    }
  }
}

/** Directory -> single RegistrationNumber, built from already-linked document paths (Tier 3.5). */
export function buildFolderRegoIndex(linkedDocs: Array<{ filePath: string; rego: string }>): Map<string, string> {
  const folderRegos = new Map<string, Set<string>>();

  for (const { filePath, rego } of linkedDocs) {
    let dir: string | null = path.dirname(filePath);
    for (let depth = 0; depth < 6 && dir !== null && dir !== path.dirname(dir); depth++, dir = path.dirname(dir)) {
      const key = dir.toUpperCase();
      if (!folderRegos.has(key)) folderRegos.set(key, new Set());
      folderRegos.get(key)!.add(rego.toUpperCase());
    }
  }

  const result = new Map<string, string>();
  for (const [dir, regos] of folderRegos) {
    if (regos.size === 1) result.set(dir, [...regos][0]);
  }
  return result;
}

export function matchFile(
  filePath: string,
  caravans: CaravanForMatching[],
  existingLink: { rego: string } | null,
  folderRegoIndex: Map<string, string>,
): FileMatch {
  const fileName = path.parse(filePath).name.toUpperCase();
  const folderPath = path.dirname(filePath).toUpperCase();
  const fileNameWithExt = path.basename(filePath);
  const fileExtension = path.extname(filePath).toLowerCase();
  const fileSizeBytes = fs.existsSync(filePath) ? fs.statSync(filePath).size : 0;

  if (existingLink) {
    const linked = caravans.find((c) => c.registrationNumber === existingLink.rego);
    return {
      filePath,
      fileName: fileNameWithExt,
      fileExtension,
      fileSizeBytes,
      confidence: 1.0,
      matchMethod: "AlreadyLinked",
      suggestedRegistrationNumber: existingLink.rego,
      suggestedCaravanDescription: linked ? describeCaravan(linked) : null,
      alreadyLinked: true,
    };
  }

  const best = {
    confidence: 0,
    suggestedRegistrationNumber: null as string | null,
    suggestedCaravanDescription: null as string | null,
    matchMethod: "",
  };

  for (const caravan of caravans) {
    tryMatch(caravan, fileName, folderPath, best);
  }

  // Tier 3.5: same vehicle-folder as an already-linked document
  if (best.confidence < 0.68) {
    let dir: string | null = path.dirname(filePath);
    for (let depth = 0; depth < 5 && dir !== null; depth++, dir = path.dirname(dir)) {
      const inferredRego = folderRegoIndex.get(dir.toUpperCase());
      if (inferredRego) {
        const c = caravans.find((x) => x.registrationNumber === inferredRego);
        const conf = depth <= 1 ? 0.68 : 0.62;
        if (conf > best.confidence) {
          best.suggestedRegistrationNumber = inferredRego;
          best.suggestedCaravanDescription = c ? describeCaravan(c) : inferredRego;
          best.confidence = conf;
          best.matchMethod = "SameFolderInference";
        }
        break;
      }
      if (path.dirname(dir) === dir) break; // reached filesystem root
    }
  }

  return {
    filePath,
    fileName: fileNameWithExt,
    fileExtension,
    fileSizeBytes,
    confidence: best.confidence,
    matchMethod: best.matchMethod,
    suggestedRegistrationNumber: best.suggestedRegistrationNumber,
    suggestedCaravanDescription: best.suggestedCaravanDescription,
    alreadyLinked: false,
  };
}
